src/main/scala-2.13+/scala/concurrent/stm/skel/AtomicArray.scala

```diff
@@ -0,0 +1,281 @@
+import scala.reflect.ClassTag
+
+
+/** `AtomicArray` implements a fixed-length indexed sequence where reads and
+ *  writes have volatile semantics.  In addition, it adds an atomic swap
+ *  operation (`swap`) and an atomic compare-and-swap (`compareAndSet`).
+ *  The collection is backed by one of the Java atomic array classes, with the
+ *  best match chosen at construction time using a manifest.
+ *
+ *  Instances of `AtomicArray[T]` are backed by `AtomicIntegerArray` if `T` is
+ *  a primitive of at most 32 bits (smaller values are padded rather than
+ *  packed).  `AtomicArray[Long]` and `AtomicArray[Double]` are backed by
+ *  `AtomicLongArray`.  All other instances of `AtomicArray[T]` are backed by
+ *  `AtomicReferenceArray` (except for `AtomicArray[Unit]`).  Floats and
+ *  doubles are stored using their raw bit representation.
+ *
+ *  This class is used in the implementation of the reference STM
+ *  implementation, but it is standalone and may be generally useful.
+ *
+ *  @author Nathan Bronson
+ */
+abstract class AtomicArray[T] extends mutable.IndexedSeq[T] with mutable.IndexedSeqOps[T, AtomicArray, AtomicArray[T]] {
+
+  // We choose to store Boolean-s (and other small primitives) each in their
+  // own Int.  This wastes space.  Another option would be to pack values into
+  // the elements of the underlying AtomicIntegerArray.  This would save space,
+  // but would require a compareAndSet loop to implement update, which adds
+  // complexity and would require some sort of back-off scheme to avoid
+  // live-lock.  A third option would be to use sun.misc.Unsafe directly, in
+  // which case volatile reads and writes of byte-sized memory locations can be
+  // performed directly.  compareAndSet of bytes would have to be emulated by
+  // integer-sized CAS.
+
+  override val iterableFactory: SeqFactory[AtomicArray] = AtomicArray.untagged
+
+  /** The length of the array */
+  def length: Int
+
+  /** The element at given index, with volatile read semantics */
+  def apply(index: Int): T
+
+  /** Update element at given index, with volatile write semantics */
+  def update(index: Int, elem: T): Unit
+
+  /** Atomic swap of the element at index */
+  def swap(index: Int, elem: T): T
+
+  /** Returns true iff previous value was expected, elem installed */
+  def compareAndSet(index: Int, expected: T, elem: T): Boolean
+
+  /** Retries compareAndSet until success, using f, then returns the old value */
+  @tailrec
+  final def getAndTransform(index: Int)(f: T => T): T = {
+    val before = apply(index)
+    if (compareAndSet(index, before, f(before))) before else getAndTransform(index)(f)
+  }
+
+  override def className = "AtomicArray"
+  
+  /** Clones this object, including the underlying Array. */
+  override def clone: AtomicArray[T] = {
+    val b = newBuilder
+    b.sizeHint(length)
+    b ++= this
+    b.result()
+  }
+
+  def newBuilder: AtomicArrayBuilder[T] = throw new AbstractMethodError
```

Comment:
Maybe we should also override `newSpecificBuilder` to throw an exception and then make `newBuilder` an alias to it? I’m not sure to understand why this method is overridden like that in 2.12, though.